### PR TITLE
[vcpkg metrics] Allow someone to opt out after build

### PR DESCRIPTION
**Pull request by strega-nil**
This PR adds an option `--disable-metrics` and an environment variable `VCPKG_DISABLE_METRICS` which turn off metrics, even if the binary was built with metrics enabled.

Additionally, as a drive-by, we now use the built-in JSON support.

Fixes issue #11542

### CODE REVIEW

File: toolsrc/src/vcpkg/metrics.cpp [hunk 1]
```diff
@@ -29,88 +30,76 @@ namespace vcpkg::Metrics
...
-    static const std::string& get_session_id()
+    // UUID format version 4, variant 1
+    // http://en.wikipedia.org/wiki/Universally_unique_identifier
+    // [0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}
+    std::string generate_random_UUID()
     {
-        static const std::string ID = generate_random_UUID();
-        return ID;
+        std::random_device rnd{};
+        std::mt19937_64 rng{rnd()};
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/metrics.cpp:93]: Preexisting: There are only 2^32 possible outputs from this; it's certainly not a UUID. Maybe rename to 'fake UUID' or something?

File: toolsrc/src/vcpkg/vcpkgcmdarguments.cpp [hunk 2]
```diff
@@ -222,6 +227,11 @@ namespace vcpkg
                     parse_switch(false, "printmetrics", args.printmetrics);
                     continue;
                 }
+                if (arg == "--no-disable-metrics")
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/vcpkgcmdarguments.cpp:230]: Not really sure why we need the switch that asks for the default...

File: toolsrc/src/vcpkg/base/json.cpp [hunk 3]
```diff
@@ -5,6 +5,8 @@
 #include <vcpkg/base/system.debug.h>
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/base/json.cpp:5]: I think all these json things should be a different PR so that they result in a different commit.

File: toolsrc/src/vcpkg/metrics.cpp (hunk 1 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/metrics.cpp:93]: sorry, I'll just remove the `mt19937_64` line.

File: toolsrc/src/vcpkg/vcpkgcmdarguments.cpp (hunk 2 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/vcpkgcmdarguments.cpp:230]: If `VCPKG_DISABLE_METRICS` is set, this switch enables the metrics again.

File: toolsrc/src/vcpkg/base/json.cpp (hunk 3 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/base/json.cpp:5]: That would be nice but they are in this commit because this is the commit that needs them.

Comment (BillyONeal) [reply toolsrc/src/vcpkg/base/json.cpp:5]: Maybe I'm blind but I don't see anything json related to any of the metrics changes.

File: toolsrc/src/vcpkg/vcpkgcmdarguments.cpp (hunk 2 above)
Comment (BillyONeal) [reply toolsrc/src/vcpkg/vcpkgcmdarguments.cpp:230]: Sounds good.

File: toolsrc/src/vcpkg/base/json.cpp (hunk 3 above)
Comment (BillyONeal) [reply toolsrc/src/vcpkg/base/json.cpp:5]: (To be clear, the json changes look good, just unrelated)

Comment (strega-nil) [reply toolsrc/src/vcpkg/base/json.cpp:5]: Metrics require floating point in JSON (see the changes from strings to using this library)

File: toolsrc/include/vcpkg/base/json.h [hunk 4]
```diff
@@ -128,6 +134,15 @@ namespace vcpkg::Json
...
 
     public:
+        Array() = default;
+        Array(Array const&) = delete;
+        Array(Array&&) = default;
+        Array& operator=(Array const&) = delete;
+        Array& operator=(Array&&) = default;
+        ~Array() = default;
+
+        Array clone() const noexcept;
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply toolsrc/include/vcpkg/base/json.h:144]: This is preexisting so feel free to not resolve in this PR. Probably need a noexcept correctness audit on all this json stuff.

File: toolsrc/src/vcpkg/base/system.process.cpp [hunk 5]
```diff
@@ -617,6 +627,10 @@ VCPKG_MSVC_WARNING(suppress : 6335) // Leaking process information handle 'proce
         Debug::print("popen(", actual_cmd_line, ")\n");
         // Flush stdout before launching external process
         fflush(stdout);
+
+        if (vcpkg::Metrics::g_metrics.lock()->metrics_enabled()) {
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/base/system.process.cpp:631]: It seems wrong that only the POSIX but not the Windows path get this treatment?

File: toolsrc/src/vcpkg/base/system.process.cpp [hunk 6]
```diff
@@ -558,6 +564,10 @@ VCPKG_MSVC_WARNING(suppress : 6335) // Leaking process information handle 'proce
         (void)env;
         Debug::print("system(", cmd_line, ")\n");
         fflush(nullptr);
+
+        if (vcpkg::Metrics::g_metrics.lock()->metrics_enabled()) {
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/base/system.process.cpp:568]: Ditto why POSIX and not Windows?

[the remark this answers is not in the document]
Comment (strega-nil) [reply toolsrc/src/vcpkg/base/system.process.cpp]: Yep!

File: toolsrc/src/vcpkg/base/system.process.cpp (hunk 5 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/base/system.process.cpp]: Windows gets this done in our environment building code.

File: toolsrc/src/vcpkg/base/json.cpp [hunk 7]
```diff
@@ -155,6 +173,7 @@ namespace vcpkg::Json
         {
             case ValueKind::Null: return Value::null(nullptr);
             case ValueKind::Boolean: return Value::boolean(boolean());
+            case ValueKind::Integer: return Value::integer(integer());
             case ValueKind::Number: return Value::number(number());
             case ValueKind::String: return Value::string(string());
             case ValueKind::Array: return Value::array(array().clone());
```
[the remark this answers is not in the document]
Comment (strega-nil) [reply toolsrc/src/vcpkg/base/json.cpp:183]: how?

[the remark this answers is not in the document]
Comment (BillyONeal) [reply toolsrc/src/vcpkg/base/json.cpp:183]: It's cloning an array (producing a temporary), then returning a string view pointing to that temporary.

[the remark this answers is not in the document]
Comment (BillyONeal) [reply toolsrc/src/vcpkg/base/json.cpp:183]: Oh, no it isn't, GitHub's code review tool is just TERRIBLE and made it look like this function started on line 119.

**Review by BillyONeal — APPROVED** [review]

**Review by dan-shaw — APPROVED** [review]

File: toolsrc/src/vcpkg/base/json.cpp [hunk 8]
```diff
@@ -865,130 +906,102 @@ namespace vcpkg::Json
...
+                    std::make_pair(0x005C, R"(\\)")  // REVERSE SOLIDUS
+                };
+                // 1. Let product be the String value consisting solely of the code unit 0x0022 (QUOTATION MARK).
+                buffer.push_back('"');
+
+                // 2. For each code point C in ! UTF16DecodeString(value), do
+                // (note that we use utf8 instead of utf16)
+                for (auto code_point : Unicode::Utf8Decoder(sv.begin(), sv.end()))
+                {
+                    bool matched = false; // early exit boolean
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/base/json.cpp:962]: 99.9% of the time when one needs a double exit bool like this that says we need extract function.

Proposal:

```
// a. If C is listed in the "Code Point" column of Table 66, then
const auto match = std::find_if(escape_sequences.begin(), escape_sequences.end(),
    [code_point](const std::pair<char32_t, const char*>& attempt) {
        return attempt.first == code_point;
    });
if (match != escape_sequences.end()) {
    // i. Set product to the string-concatenation of product and the escape sequence for C as
    // specified in the "Escape Sequence" column of the corresponding row.
    buffer.append(match->second);
    break;
}
```

File: toolsrc/src/vcpkg/base/json.cpp [hunk 9]
```diff
@@ -865,130 +906,102 @@ namespace vcpkg::Json
...
 
-        // 4. Return product.
-        return product;
-    }
+            // taken from the ECMAScript 2020 standard, 24.5.2.2: Runtime Semantics: QuoteJSONString
+            void append_quoted_json_string(StringView sv)
+            {
+                // Table 66: JSON Single Character Escape Sequences
+                constexpr static std::array<std::pair<char32_t, const char*>, 7> escape_sequences = {
+                    std::make_pair(0x0008, R"(\b)"), // BACKSPACE
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/base/json.cpp:947]: Is make_pair really needed here or would {} suffice? (Honestly asking)

**Review by BillyONeal — APPROVED** [review]

File: toolsrc/src/vcpkg/base/json.cpp (hunk 9 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/base/json.cpp:947]: `{}` would probably suffice.

File: toolsrc/src/vcpkg/base/json.cpp (hunk 8 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/base/json.cpp:962]: Hmm, good point.

**Comment by dan-shaw** [comment]
LGTM

**Comment by strega-nil** [comment]
@dan-shaw could you message me on teams?
